### DISCUSSION TITLE
Custom Extension with Multiple Types

### DISCUSSION THREAD

**Discussion by ChappIO** [directus/directus#17454] [Archive]
Here I am again (sorry).

I am currently working on an extension for Directus which exposes metrics for prometheus (our central metrics database). To make this work I would need hooks to collect request metrics as well as endpoint to expose the metrics. However, right now an extension has to be one of the two.

Is this problem uniquely mine?

**Comment by ChappIO** [comment]
I currently solved this by not registering a custom endpoint but instead just building my endpoint as an express middleware during the `middlewares.after` init hook:

```javascript
init('middlewares.after', async ({app}) => {
  app.use('/my-custom-endpoint', (req, res, next) => {
    // do things
  });
});
```

I suppose that is a workaround I can live with.

**Comment by paescuj** [comment]
You're probably looking for the [`bundle` extension type](https://github.com/directus/directus/pull/15672), added in [v9.21.0](https://github.com/directus/directus/releases/tag/v9.21.0).

**Reply by ChappIO** [reply]
Interesting. What does this type do? It doesn't seem to be documented and from the issue description it just sounds like any other package.json which pulls in multiple extensions. 

When I load multiple extensions, will they be able to interact with (or import) eachother?

**Reply by rijkvanzanten** [reply]
@ChappIO The main difference is that it allows you to create multiple separate extensions, that are all bundled into the same dist files. That way, it's able to de-duplicate packages and other shared code.

**Reply by Nitwel** [reply]
Bundles multiple extensions into a single one. When building you end up with a single file containing all extensions in it. A great benefit of bundles are that you can share dependencies between them.

Extensions can interact with another without the need for bundles though. But you can share your logic between them inside a bundle if that is what you mean.

**Reply by ChappIO** [reply]
Then I think I misunderstood the issue description. 

You're saying that the bundle type is NOT a package.json that just includes other npm packages (extensions) but rather a single package that can provide multiple extensions?

**Reply by paescuj** [reply]
Maybe this helps 😃 

[![asciicast](https://asciinema.org/a/558356.svg)](https://asciinema.org/a/558356)

**Reply by ChappIO** [reply]
> Maybe this helps 😃 
> 


It does 😊 I will try some thing with this soon. Thank you!

**Reply by ernst86913** [reply]
Cool... is there already some documentation available for bundle extensions? maybe in draft state?

**Reply by paescuj** [reply]
Not yet, but it's in the planning. Keep an eye on https://github.com/directus/docs/issues/261.

**Reply by ChappIO** [reply]
@Nitwel you say extensions can already interact with eachother. However, I cannot use both a default export for my extension as well as provide functions for them to call.

In my case, I have an instance of a class where I want people to register more metrics from their extension. How could I provide this instance to them?

Ideally they would do something like this:

```
import {getRegister} from "directus-extension-prometheus";

export default defineHook(({init}, {services, getSchema, database, env}) => {
  const register = getRegister();
  
  // register their own metrics
});
```

Currently when I add a (non default) export to my index, Rollup will complain about it:

```
Entry module "src/index.ts" is using named and default exports together. Consumers of your bundle will have to use `chunk.default` to access the default export, which may not be what you want.
```

This is caused because my extension has both a default export and exports the `getRegister` function:

```typescript

export function getRegister() {
 // ...
 }
 
 export default defineHook(...);
```

Basically, my problem: I would like other extensions to have access to my `register` object.

**Reply by rijkvanzanten** [reply]
@ChappIO I think that particular warning stems from ESM vs CJS in the node context. Extensions are currently built to CJS, but your code uses ESM style exports, which is then transpiled by rollup back to CJS. That's how you end up with effectively:

```js
module.exports = {
  default: x
}
```

which might be unexpected, as you then have to do `require('./file').default` to get to the "default" export. (CJS->ESM is really the Python 2 to 3 of the Node.js ecosystem 😞 )

**Reply by ChappIO** [reply]
@rijkvanzanten i am using defaults using directus-extension-sdk and typescript though. So I don't think I have access to the rollup configuration to change that right?

**Reply by rijkvanzanten** [reply]
It's also not really something that can be fixed in config, as `CJS` itself doesn't have the notion of "Default" exports like ESM does 🤔

**Reply by ChappIO** [reply]
@rijkvanzanten Maybe I should clarify that I use typescript. So compiling TS to CJS should work right?

Anyway, I've converted my library to CJS style exports and it works fine now. Just feels strange to use that syntax in TS 😂  but it works.

Thanks again for this cool project. I have never prototyped ideas as fast as this before.

**Reply by rijkvanzanten** [reply]
> Maybe I should clarify that I use typescript. So compiling TS to CJS should work right?

On paper, yes 👀  TypeScript is a special flavor of annoying, as it uses ESM _syntax_ (eg the `import x from 'y'`), but then transpiles that to "straight" CJS, which makes things confusing, as that also means that it's legal to import files without file extensions etc. Anywho.....

> Anyway, I've converted my library to CJS style exports and it works fine now. Just feels strange to use that syntax in TS 😂 but it works.

Glad to hear it! And yes I agree 😵 The longer-term plan is to make the whole codebase (including the API) ESM only to hopefully get rid of this mess, but that's an ongoing effort.

> Thanks again for this cool project. I have never prototyped ideas as fast as this before.

Sweet! Glad you like it ☺️
